File: src/libs/m3/net/DataQueue.cc
/*
 * Copyright (C) 2019-2021 Nils Asmussen, Barkhausen Institut
 * Copyright (C) 2018, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/net/DataQueue.h>

namespace m3 {

DataQueue::Item::Item(NetEventChannel::DataMessage const *msg,
                      NetEventChannel::Event &&event) noexcept
    : _msg(msg), _event(std::move(event)), _pos(0) {
}

IpAddr DataQueue::Item::src_addr() const noexcept {
    return IpAddr(_msg->addr);
}

port_t DataQueue::Item::src_port() const noexcept {
    return _msg->port;
}

const uchar *DataQueue::Item::get_data() const noexcept {
    return _msg->data;
}

size_t DataQueue::Item::get_size() const noexcept {
    return _msg->size;
}

size_t DataQueue::Item::get_pos() const noexcept {
    return _pos;
}

void DataQueue::Item::set_pos(size_t pos) noexcept {
    assert(pos <= get_size());
    _pos = pos;
}

DataQueue::~DataQueue() {
    clear();
}

void DataQueue::append(Item *item) noexcept {
    _recv_queue.append(item);
}

bool DataQueue::has_data() const noexcept {
    return _recv_queue.length() > 0;
}

bool DataQueue::get_next_data(const uchar **data, size_t *size, Endpoint *ep) noexcept {
    if(!has_data())
        return false;

    Item &item = *_recv_queue.begin();
    *data = item.get_data() + item.get_pos();
    *size = item.get_size() - item.get_pos();
    if(ep)
        *ep = Endpoint(item.src_addr(), item.src_port());
    return true;
}

void DataQueue::ack_data(size_t size) noexcept {
    // May be called exactly once for every successful invocation of get_next_data().
    assert(_recv_queue.length() > 0);

    Item &item = *_recv_queue.begin();
    assert(item.get_pos() + size <= item.get_size());
    item.set_pos(item.get_pos() + size);

    // Remove item if its data is exhausted
    if(item.get_pos() >= item.get_size())
        delete _recv_queue.remove_first();
}

void DataQueue::clear() noexcept {
    Item *item;
    while((item = _recv_queue.remove_first()) != nullptr)
        delete item;
}

}

File: src/libs/rust/m3/src/net/dataqueue.rs
/*
 * Copyright (C) 2021 Nils Asmussen, Barkhausen Institut
 * Copyright (C) 2017, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use core::cmp;

use crate::col::DList;
use crate::net::{event, Endpoint, IpAddr, NetEvent, Port};

struct Item {
    event: NetEvent,
    pos: usize,
}

impl Item {
    fn data(&self) -> &[u8] {
        &self.msg().data[self.pos..self.size()]
    }

    fn size(&self) -> usize {
        self.msg().size as usize
    }

    fn addr(&self) -> IpAddr {
        IpAddr(self.msg().addr as u32)
    }

    fn port(&self) -> Port {
        self.msg().port as Port
    }

    fn msg(&self) -> &event::DataMessage {
        self.event.msg::<event::DataMessage>()
    }
}

#[derive(Default)]
pub struct DataQueue {
    items: DList<Item>,
}

impl DataQueue {
    pub fn append(&mut self, event: NetEvent, pos: usize) {
        self.items.push_back(Item { event, pos });
    }

    pub fn clear(&mut self) {
        self.items.clear();
    }

    pub fn has_data(&self) -> bool {
        !self.items.is_empty()
    }

    pub fn next_data<F, R>(&mut self, len: usize, consume: &mut F) -> Option<(usize, R)>
    where
        F: FnMut(&[u8], Endpoint) -> (usize, R),
    {
        if let Some(first) = self.items.front_mut() {
            let data = first.data();
            let amount = cmp::min(len, data.len());
            let ep = Endpoint::new(first.addr(), first.port());
            let (amount, res) = consume(&data[0..amount], ep);
            if amount >= data.len() {
                self.items.pop_front();
            }
            else if amount == 0 {
                return None;
            }
            else {
                first.pos += amount;
            }
            Some((amount, res))
        }
        else {
            None
        }
    }
}
